source/common/http/filter_manager.cc

```diff
@@ -1044,9 +1060,55 @@ void FilterManager::encodeHeaders(ActiveStreamEncoderFilter* filter, ResponseHea
     }
   }
 
-  const bool modified_end_stream = (end_stream && continue_data_entry == encoder_filters_.end());
+  bool modified_end_stream = (end_stream && continue_data_entry == encoder_filters_.end());
+
+  // Capture the original modified_end_stream value. We may modify it again below if we add
+  // a response body.
+  const bool original_modified_end_stream = modified_end_stream;
+
+  // See if we should do an upstream rewrite. For local replies, always try to do a rewrite.
+  // For upstream replies, only try to do a rewrite if some filter rule actually applies.
+  // The rewrite config supports unconditional rewrites with no filter rules. This makes sense
+  // for local reply rewrites (e.g. one consistent format for all responses generated at Envoy)
+  // but makes a lot less sense for all responses upstream.
+  state_.do_response_rewrite_ =
+      local_reply_data_ != nullptr ||
+      local_reply_.matchesAnyMapper(filter_manager_callbacks_.requestHeaders().ptr(),
+                                    *filter_manager_callbacks_.responseHeaders(),
+                                    filter_manager_callbacks_.responseTrailers().ptr(),
+                                    stream_info_);
+  ENVOY_STREAM_LOG(debug,
+                   "FilterManager::encodeHeaders: end_stream={}, modified_end_stream={}, "
+                   "state_.do_response_rewrite_={}",
+                   *this, end_stream, modified_end_stream, state_.do_response_rewrite_);
+
+  if (state_.do_response_rewrite_) {
+    // Try to rewrite the response. This may end up not doing any actual rewrite if this is
+    // a local reply and there is no matching rule. If it's not a local reply, we know there's
+    // a match at this point since that's a condition for state_.do_response_rewrite_ above.
+    rewriteResponse();
```

Comment:
At this point in the code, we've already run through the filter chain.

Question: should we do response rewriting before or after the filter chain? If we do it before, then the filter chain has an opportunity to see the rewritten response and make its own changes. If we do it after, then we give response rewriting the final say and the highest precedence.

Reply:
As discussed offline, if we can tell ahead of time we're doing a rewrite, I think we want to let the filters see the reply that is going to the client.  If they do their own stats tracking of say "how many 500s happened" I think they want to see what will be written.
@snowp  @lizan can I get a non-googly take on this question?

Reply:
One interesting case to consider is when a filter itself returns a status (or sets a header) that would match a rewrite rule. For example, if ext authz returned 403, the user may want the same 403 rewrite rule to apply as it would have if an upstream service had returned it. Same story might reply to a sentinel header like `x-my-custom-failure-header` that could get set by your upstream services or by a custom filter. Either way the user might want the same rewrite rules to match/apply.

Reply:
Yeah this is a tricky one. I haven't yet thought about this a huge amount, but I think in a perfect world we would do something like the following:
1) Unconditional rewriting happens before the encoder filter chain runs.
2) Conditional rewriting (matching) actually is attempted after *each* encoder filter runs. This I think is the behavior we want where it allows each filter to change something and then possibly have a transform happen before the next filter runs.

I do wonder if somehow this could be built using the new matching infra that @snowp created but I haven't looked at it in detail yet.

Let me know if it would be easier to chat about this in a short meeting or in a doc.

Reply:
yeah, that matched my thought, where (I think) this one qualifies as unconditional and would happen first.